### DISCUSSION TITLE
SSR Roadmap

### DISCUSSION THREAD

**Discussion by matthewp** [withastro/roadmap#70] [Proposal]
With https://github.com/withastro/rfcs/pull/46 accepted, we are now in a place where our build process matches requirements for SSR. Given that, now is a great opportunity to establish the roadmap for releasing SSR.

I see this approximate order (no particular timeline given for each)

1. __Static build__
    - Stabilize `--experimental-static-build` so that it works with all examples / typical usage.
    - Confirm limits to number of pages that can be built. Optimize where we can.
1. __SSR experimentation__
    - Add a low-level implementation of SSR behind a flag. (`--experimental-ssr`)
      - Use the same mechanisms as the static build, but doesn't prerender.
      - Probably needs the *router* to be included in the build.
      - Only Node.js support at this point.
1. __SSR features__
    - Quickly iterate with the community on features needed in an SSR environment.
    - Access to query params
    - Sessions / login
1. __Static build release and more SSR__
    - Flip `astro build` so the static build is default and `--legacy-build` flag enables the scanning-based build.
    - Continue working on SSR features
1. __Finalize SSR APIs__
    - APIs for creating adapters/plugins/connectors to specific providers.
    - RFCs for stuff like `Astro.request` (maybe one big RFC?)
1. __SSR support for hosts__
    - Remove experimental flag.
    - - Figure out how build should work (ssr with a flag? unflagged)?
    - Launch with support for one big host.

**Comment by RafidMuhymin** [comment]
@matthewp Can you explain the points below that you have mentioned above??

> Use the same mechanisms as the static build, but doesn't pretender

> Probably needs the router to be included in the build

**Comment by matthewp** [comment]
> Use the same mechanisms as the static build, but doesn't pretender

For SSR we need to compile your site to JavaScript, this means compiling all .astro pages and their dependencies. This is because we cannot run Vite in production (on a server).

The static build works this way. The only difference is that the static build renders all of the pages to .html files, and SSR will not (it will render on demand).

> Probably needs the router to be included in the build

In dev mode Astro has a router that matches a URL to the page that it should render. This needs to be included in SSR so that you can route your server requests.

**Comment by boehs** [comment]
I oppose ~~SSR features~~ A full backend framework in astro. Unix philosophy says "do one thing and do it well". Astro is a great static site generator, and I don't think SSR features will benefit it.

* It will likely never reach feature parity with backend frameworks like express
* Attempts to do so will result in time spent on bettering the framework instead on reinventing the wheel

I instead encourage discussion of how we can improve the experience of people using backend frameworks+templating, I plan to open a RFC discussion on this topic shortly. Astro is in a unique position for use in templating, it could be the first framework to support it fully. I really want to see it go that way.

As a closing note, I used to use sveltekit, I switched away from svelte to astro exactly so I could regain control.

**Reply by matthewp** [reply]
Thanks for the feedback. The intention here is to allow Astro to be used as a templating language on the server. It would not replace the functionality of something like Express, and you could indeed use Express, including its routing, with Astro sitting on top.

**Reply by FredKSchott** [reply]
Thanks for the feedback! It's also worth calling out that this work will make Astro a better static build tool as well. The work that @matthewp has done to explore SSR has already increased our build performance by something like 75% iirc with improved static builds.

**Reply by boehs** [reply]
> Thanks for the feedback. The intention here is to allow Astro to be used as a templating language on the server. It would not replace the functionality of something like Express, and you could indeed use Express, including its routing, with Astro sitting on top.

~~I'm sorry, I misunderstood! I thought it was something like sveltekit, where I found myself frustrated by the lock in~~. Is there a syntax example of this somewhere? ~~Very interested to see how this works with backend frameworks.~~

---

> Edit: Oh, I understand now. Express -> astro please render page -> ok here is page all done!

That's quite interesting! Would you guys be opposed to also supporting other templating languages for a roundabout workflow that looks like

> Astro -> Static build -> Some template files that can be filled by eta, handlebars, ejs, etc -> framework?

I would like to draft a RFC that looks like this and maybe even implement it, because I am still attached to my static files for some reasons. I would be excited to chat with matt about what this would look like!

---

Also, very very sorry if it came across as aggressive, I have been loving astro and it was not my intention whatsoever.

**Comment by isaac-mcfadyen** [comment]
Looks good! I see the note on their at Stage 2 about only supporting Node.js. Any thoughts about which stage Node.js will not be a requirement? I'm currently using Cloudflare Workers which [don't have a Node.js runtime](https://developers.cloudflare.com/workers/learning/how-workers-works) and would prefer to stick with them over something like Vercel/Netlify Functions as they're much faster than Node.

**Reply by matthewp** [reply]
Cloudflare will definitely be supported. The (Finalize SSR APIs) step was meant to encompass creating adapters (or adapter APIs so other can write the actual adapters) for major hosts and that will include Cloudflare.

We could have support earlier than this if people contribute though! At the earlier stages there won't be an *easy* way to deploy to a host, essentially we will just be building the files and you'll be responsible for deploying it to your host and integrating it with routing, etc. 

The reason why we said Node.js-only at the start is because the build output assumes a node_modules environment. But if we change the build to bundle everything together, that won't be necessary.

**Comment by buzzeins** [comment]
@matthewp Is there some example for using Astro as a templating engine, for example in Express, Fastify or Nestjs? I'd really appreciate 👍 

>The intention here is to allow **Astro to be used as a templating language on the server**. It would not replace the functionality of something like Express, **and you could indeed use Express, including its routing, with Astro sitting on top**.

**Reply by matthewp** [reply]
Not yet, I'm working on an SSR flag this week and next and then can hopefully give you an example.

**Reply by buzzeins** [reply]
Thanks, SSR will definitely be a bigdeal for Astro :)

**Comment by wassfila** [comment]
the future is serverless, and Astro's SSR is a pioneer in shifting from static CDN content to taking advantage of the Edge !
That is not shadowing Express, we're talking middleware.

**Comment by kireerik** [comment]
Is this roadmap proposal still relevant?

**Reply by Princesseuh** [reply]
No, SSR has been shipped for a long time now and all the points in the roadmap highlighted there have been done.
